services/addons/images/firmware_manager/firmware_manager.py

```diff
@@ -259,18 +299,67 @@ def check_for_upgrades():
             ):
                 continue
 
+            # Check if latest ping was unsuccessful
+            if not was_latest_ping_successful_for_rsu(rsu["ipv4_address"]):
+                logging.info(
+                    f"Skipping firmware upgrade for '{rsu['ipv4_address']}': device is unreachable"
+                )
+                continue
+
             # Add the RSU to the upgrade queue and record the necessary upgrade information
             logging.info(
                 f"Adding '{rsu["ipv4_address"]}' to the firmware manager upgrade queue"
             )
             upgrade_queue.extend([rsu["ipv4_address"]])
             upgrade_queue_info[rsu["ipv4_address"]] = rsu
-            logging.info(f"Firmware upgrade successfully started for '{rsu["ipv4_address"]}'")
+            logging.info(
+                f"Firmware upgrade successfully started for '{rsu["ipv4_address"]}'"
+            )
 
         # Start any processes that can be started
         start_tasks_from_queue()
 
 
+def was_latest_ping_successful_for_rsu(rsu_ip):
+    query = f"select result from ping where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}') order by timestamp desc limit 1"
+    latest_ping_successful = pgquery.query_db(query)[0][0]

```

Comment:
issue: if we receive an empty array from `query_db()` we're at risk of getting and index out of bound exception.  We should first check the length of the results, then try to access the data within them. So something like the following would work
```suggestion
    pings = pgquery.query_db(query)[0][0]
    if len(pings) == 0 or len(pings[0]) == 0:
        return False
    latest_ping_successful = pings[0][0]
```

Reply:
Addressed!